File: install/identity-qr-reader/src/redeem.test.ts
import { describe, expect, it, vi } from "vitest";
import { formatRedeemLog, redeemQrUrl } from "./redeem.js";

describe("redeemQrUrl", () => {
  it("posts qrUrl to Identity bridge redeem", async () => {
    const fetchMock = vi.fn().mockResolvedValue({
      ok: true,
      json: async () => ({
        ok: true,
        validationId: "val_1",
        memberId: "mem_1",
        correlationOutcome: "AUTHORIZED",
      }),
    });

    const result = await redeemQrUrl(
      {
        identityUrl: "http://localhost:3100",
        deviceKey: "idb_test_key",
        emitDetection: true,
      },
      "http://localhost:3100/r/intent1?t=secret",
      fetchMock as typeof fetch,
    );

    expect(result.ok).toBe(true);
    expect(fetchMock).toHaveBeenCalledWith(
      "http://localhost:3100/api/bridge/intent/redeem",
      expect.objectContaining({
        method: "POST",
        headers: expect.objectContaining({
          Authorization: "Bearer idb_test_key",
        }),
      }),
    );
    const body = JSON.parse(String(fetchMock.mock.calls[0]?.[1]?.body));
    expect(body.qrUrl).toContain("/r/intent1");
    expect(body.emitDetection).toBe(true);
  });

  it("formats error responses in Portuguese", () => {
    const line = formatRedeemLog({
      ok: false,
      status: 403,
      data: { error: "QR expirado.", code: "INTENT_EXPIRED" },
    });
    expect(line).toContain("ERRO 403");
    expect(line).toContain("INTENT_EXPIRED");
  });
});

File: install/qr-reader-agent/internal/redeem/redeem_test.go
package redeem

import (
	"context"
	"io"
	"net/http"
	"strings"
	"testing"
)

type roundTripFunc func(req *http.Request) (*http.Response, error)

func (f roundTripFunc) RoundTrip(req *http.Request) (*http.Response, error) {
	return f(req)
}

func TestRedeemQRURL(t *testing.T) {
	var auth string
	var body string
	client := &http.Client{
		Transport: roundTripFunc(func(req *http.Request) (*http.Response, error) {
			auth = req.Header.Get("Authorization")
			raw, _ := io.ReadAll(req.Body)
			body = string(raw)
			return &http.Response{
				StatusCode: 200,
				Body:       io.NopCloser(strings.NewReader(`{"ok":true,"validationId":"val_1","memberId":"mem_1","correlationOutcome":"AUTHORIZED"}`)),
				Header:     make(http.Header),
			}, nil
		}),
	}

	c := NewClient(ClientConfig{
		IdentityURL:   "http://localhost:3100",
		DeviceKey:     "idb_test_key",
		EmitDetection: true,
	}, client)

	result := c.RedeemQRURL(context.Background(), "http://localhost:3100/r/intent1?t=secret")
	if !result.OK {
		t.Fatalf("expected ok result: %+v", result)
	}
	if auth != "Bearer idb_test_key" {
		t.Fatalf("unexpected auth: %s", auth)
	}
	if !strings.Contains(body, `"emitDetection":true`) || !strings.Contains(body, "/r/intent1") {
		t.Fatalf("unexpected body: %s", body)
	}
}

func TestFormatLogError(t *testing.T) {
	line := FormatLog(Result{
		OK:     false,
		Status: 403,
		Data:   Response{Error: "QR expirado.", Code: "INTENT_EXPIRED"},
	})
	if !strings.Contains(line, "ERRO 403") || !strings.Contains(line, "INTENT_EXPIRED") {
		t.Fatalf("unexpected line: %s", line)
	}
}
